codelab/01_kpt-validate/tutorial.md

```diff
@@ -0,0 +1,259 @@
+# Validate Your Configurations in skaffold Workflow
+
+<walkthrough-disable-features toc></walkthrough-disable-features>
+
+## Introduction
+
+#### What is kpt?
+
+Kpt is [an OSS tool](https://github.com/GoogleContainerTools/kpt) for Kubernetes packaging, which uses a standard format to bundle, publish, customize, update, and apply configuration manifests.
+
+#### What kpt can help you
+    
+*  You will get an hands-on off-the-shelf experience about the **GitOps** CI/CD workflow in skaffold.
+*  You can validate each of your config changes **declaratively**.
+*  You **won't** encounter **version conflict** if the config hydration (a.k.a kustomize) mismatch with the deployment tool (e.g. kubectl). 
+*  You can prune your resources accurately with [a three-way merge strategy](https://kubectl.docs.kubernetes.io/pages/app_management/field_merge_semantics.html). 
+
+#### What you'll learn
+    
+*  How to add a validation to your config changes. 
+*  How to define validation rules in the form of a declarative configuration.
+*  How to use the validation in kustomized resources. 
+
+## Prerequisites
+
+If you are new to skaffold, you can check out [the skaffold tutorials](https://skaffold.dev/docs/tutorials/) to get a basic idea. 
+Or just follow this codelab, we will explain what happens in each step.
+
+#### Install `skaffold`
+
+*   Check if `skaffold` is installed and the installed version is >= v1.17.0
+    ```bash
+    skaffold version
+    ```
+
+*   If you haven't installed `skaffold` previously, run
+    ```bash
+    curl -Lo skaffold https://storage.googleapis.com/skaffold/builds/latest/skaffold-linux-amd64 && \
+    sudo install skaffold /usr/bin/ | bash
+    ```  
+*   To upgrade skaffold to a newer version, run
+    ```bash
+    sudo rm -f /usr/bin/skaffold && curl -Lo skaffold https://storage.googleapis.com/skaffold/builds/latest/skaffold-linux-amd64 && \
+    sudo install skaffold /usr/bin/ | bash
+    ```
+    
+#### Install `kpt`
+
+*   Check if `kpt` is installed and the installed version is >= 0.34.0
+    ```bash
+    kpt version
+    ```
+*   If you haven't installed `kpt` previously, run
+    ```bash
+    sudo apt-get install google-cloud-sdk-kpt
+    ```  
+    
+#### Install `kustomize`
+
+*   Check if `kustomize` is installed and the installed version is >= v3.4.3
+    ```bash
+    kustomize version
+    ```
+*   If you haven't installed `kustomize` previously, run
+    ```bash
+        curl -s "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh"  | bash &&  sudo mv kustomize /usr/local/go/bin
+    ```  
+    
+## Getting started
+
+Time to complete: **About 3 minutes**
+
+
+*   In your
+    <walkthrough-editor-spotlight spotlightId="menu-terminal-new-terminal">terminal</walkthrough-editor-spotlight>,
+    run this command to start your local minikube cluster:
+    
+    ```bash
+    minikube start
```

Comment:
should we have a section about installing minikube in the prerequisites?

Reply:
cloud code auto installs minikube. https://cloud.google.com/code/docs/vscode/adding-a-cluster